colflow: unexpectedly not all closers were closed before Flow.Cleanup

Extracted from #60824.

sql/logictest.TestLogic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2925863&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2925863&tab=artifacts#/) on master @ [05880958efb6fda9ab80ed16495d98922de3fbf6](https://github.com/cockroachdb/cockroach/commits/05880958efb6fda9ab80ed16495d98922de3fbf6):

Fatal error:

```
panic: expected 2 components to be closed, but found that only 0 were

```
Stack: 

```
goroutine 9453985 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.InternalError(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:192
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Cleanup(0xc146920aa0, 0x51d5e40, 0xc037fcf240)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:285 +0x505
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func8()
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:373 +0x82
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc0c811c600, 0x51f4140, 0xc05fff0318, 0xc0c811cae0, 0x3, 0x5243ee0, 0xc08e468c00, 0xc109496401, 0xc0b0f19078, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1055 +0x4cc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0c811c600, 0x51f4140, 0xc05fff0318, 0xc0c811cae0, 0x5243ee0, 0xc08e468c00, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:915 +0x75b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0c811c600, 0x51f4140, 0xc05fff0318, 0x5200700, 0xc02c975f90, 0xc0979d1bc0, 0x3c, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:659 +0xf86
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0c811c600, 0x51d5e40, 0xc0ce4d52c0, 0x5200700, 0xc02c975f90, 0xc0979d1bc0, 0x3c, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:122 +0xb14
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x5200700, 0xc02c975f90, 0xc0979d1bc0, 0x3c, 0x0, 0x0, 0xad5db37, 0xed8193acf, 0x0, 0xad5db37, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1527 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc0c811c600, 0x51d5e40, 0xc0ce4d52c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1529 +0x1457
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0c811c600, 0x51d5f00, 0xc08e468870, 0xc081bf7e00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1450 +0x228
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1(0xc0c811c600, 0x51d5f00, 0xc08e468870, 0xc0b0f18960, 0xc08e4689c0, 0x5188200, 0xc0ce4d5140, 0xc02ad0a630, 0xc17f462d20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:202 +0x65
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:201 +0x2f2
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLogic/fakedist-disk/serial/serial_virtual_sequence
E210426 22:50:55.106794 9444974 jobs/adopt.go:260  [n1] 954  job 653417037656260609: adoption completed with error failed to ingest index entries during backfill: duplicate key value violates unique constraint "bar"
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63262 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-20.2]
- #61657 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-21.1]
- #58269 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-20.1]
</p>
</details>
/cc @jordanlewis
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/colexec/parallel_unordered_synchronizer.go
+++ b/pkg/sql/colexec/parallel_unordered_synchronizer.go
@@ -102,6 +102,9 @@ type ParallelUnorderedSynchronizer struct {
 	bufferedMeta []execinfrapb.ProducerMetadata
 }
 
+var _ colexecop.DrainableOperator = &ParallelUnorderedSynchronizer{}
+var _ colexecop.ClosableOperator = &ParallelUnorderedSynchronizer{}
+
 // ChildCount implements the execinfra.OpNode interface.
 func (s *ParallelUnorderedSynchronizer) ChildCount(verbose bool) int {
 	return len(s.inputs)
@@ -414,4 +417,30 @@ func (s *ParallelUnorderedSynchronizer) DrainMeta() []execinfrapb.ProducerMetada
 	// Done.
 	s.setState(parallelUnorderedSynchronizerStateDone)
 	return s.bufferedMeta
+}
+
+// Close is part of the colexecop.ClosableOperator interface.
+func (s *ParallelUnorderedSynchronizer) Close(context.Context) error {
+	if state := s.getState(); state != parallelUnorderedSynchronizerStateUninitialized {
+		// Input goroutines have been started and will take care of closing the
+		// closers from the corresponding input trees, so we don't need to do
+		// anything.
+		return nil
+	}
+	// If the synchronizer is in "uninitialized" state, it means that the
+	// goroutines for each input haven't been started, so they won't be able to
+	// close the Closers from the corresponding trees. In such a scenario the
+	// synchronizer must close all of them from all input trees. Note that it is
+	// ok to close some input trees even if they haven't been initialized.
+	//
+	// Note that at this point we know that the input goroutines won't be
+	// spawned up (our consumer won't call Next/DrainMeta after calling Close),
+	// so it is safe to close all closers from this goroutine.
+	var lastErr error
+	for _, input := range s.inputs {
+		if err := input.ToClose.Close(s.Ctx); err != nil {
+			lastErr = err
+		}
+	}
+	return lastErr
 }

--- a/pkg/sql/colexec/parallel_unordered_synchronizer_test.go
+++ b/pkg/sql/colexec/parallel_unordered_synchronizer_test.go
@@ -13,6 +13,7 @@ package colexec
 import (
 	"context"
 	"fmt"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"testing"
@@ -194,6 +195,50 @@ func TestUnorderedSynchronizerNoLeaksOnError(t *testing.T) {
 	require.Zero(t, len(s.DrainMeta()))
 	// This is the crux of the test: assert that all inputs have finished.
 	require.Equal(t, len(inputs), int(atomic.LoadUint32(&s.numFinishedInputs)))
+}
+
+// TestParallelUnorderedSyncClosesInputs verifies that the parallel unordered
+// synchronizer closes the input trees if it encounters a panic during the
+// initialization.
+func TestParallelUnorderedSyncClosesInputs(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	const injectedPanicMsg = "injected panic"
+	inputs := make([]colexecargs.OpWithMetaInfo, 2)
+
+	// Create the first input that is responsible for tracking whether the
+	// closure occurred as expected.
+	closed := false
+	firstInput := &colexecop.CallbackOperator{
+		CloseCb: func(context.Context) error {
+			closed = true
+			return nil
+		},
+	}
+	inputs[0].Root = firstInput
+	inputs[0].ToClose = append(inputs[0].ToClose, firstInput)
+
+	// Create the second input that injects a panic into Init.
+	inputs[1].Root = &colexecop.CallbackOperator{
+		InitCb: func(context.Context) {
+			colexecerror.InternalError(errors.New(injectedPanicMsg))
+		},
+	}
+
+	// Create and initialize (but don't run) the synchronizer.
+	var wg sync.WaitGroup
+	s := NewParallelUnorderedSynchronizer(inputs, &wg)
+	err := colexecerror.CatchVectorizedRuntimeError(func() { s.Init(ctx) })
+	require.NotNil(t, err)
+	require.True(t, strings.Contains(err.Error(), injectedPanicMsg))
+
+	// In the production setting, the user of the synchronizer is still expected
+	// to close it, even if a panic is encountered in Init, so we do the same
+	// thing here and verify that the first input is properly closed.
+	require.NoError(t, s.Close(ctx))
+	require.True(t, closed)
 }
 
 func BenchmarkParallelUnorderedSynchronizer(b *testing.B) {

--- a/pkg/sql/colexecop/operator.go
+++ b/pkg/sql/colexecop/operator.go
@@ -138,6 +138,9 @@ type BufferingInMemoryOperator interface {
 // if we have a simple project on top of a disk-backed operator, that simple
 // project needs to implement this interface so that Close() call could be
 // propagated correctly).
+// TODO(yuzefovich): clarify the contract that Close must be safe to call even
+// if the Operator - when it implements the Closer interface - hasn't been
+// initialized.
 type Closer interface {
 	Close(ctx context.Context) error
 }

--- a/pkg/sql/colexecop/testutils.go
+++ b/pkg/sql/colexecop/testutils.go
@@ -137,19 +137,39 @@ func (s *RepeatableBatchSource) ResetBatchesToReturn(b int) {
 	s.batchesReturned = 0
 }
 
-// CallbackOperator is a testing utility struct that delegates Next calls to a
-// callback provided by the user.
+// CallbackOperator is a testing utility struct that delegates calls to Init,
+// Next, and Close to the callbacks provided by the user.
 type CallbackOperator struct {
 	ZeroInputNode
-	NextCb func() coldata.Batch
+	InitCb  func(context.Context)
+	NextCb  func() coldata.Batch
+	CloseCb func(context.Context) error
 }
 
+var _ ClosableOperator = &CallbackOperator{}
+
 // Init is part of the Operator interface.
-func (o *CallbackOperator) Init(context.Context) {}
+func (o *CallbackOperator) Init(ctx context.Context) {
+	if o.InitCb == nil {
+		return
+	}
+	o.InitCb(ctx)
+}
 
 // Next is part of the Operator interface.
 func (o *CallbackOperator) Next() coldata.Batch {
+	if o.NextCb == nil {
+		return coldata.ZeroBatch
+	}
 	return o.NextCb()
+}
+
+// Close is part of the ClosableOperator interface.
+func (o *CallbackOperator) Close(ctx context.Context) error {
+	if o.CloseCb == nil {
+		return nil
+	}
+	return o.CloseCb(ctx)
 }
 
 // TestingSemaphore is a semaphore.Semaphore that never blocks and is always

--- a/pkg/sql/colflow/vectorized_flow.go
+++ b/pkg/sql/colflow/vectorized_flow.go
@@ -896,11 +896,7 @@ func (s *vectorizedFlowCreator) setupInput(
 				opWithMetaInfo = colexecargs.OpWithMetaInfo{
 					Root:            sync,
 					MetadataSources: colexecop.MetadataSources{sync},
-					// ToClose is set to nil because the
-					// ParallelUnorderedSynchronizer takes care of closing these
-					// components itself since they need to be closed from the
-					// same goroutine as Next.
-					ToClose: nil,
+					ToClose:         colexecop.Closers{sync},
 				}
 				s.operatorConcurrency = true
 			}

--- a/pkg/sql/colflow/vectorized_flow.go
+++ b/pkg/sql/colflow/vectorized_flow.go
@@ -285,7 +285,7 @@ func (f *vectorizedFlow) Cleanup(ctx context.Context) {
 	// This cleans up all the memory and disk monitoring of the vectorized flow.
 	f.creator.cleanup(ctx)
 
-	if f.Cfg.TestingKnobs.CheckVectorizedFlowIsClosedCorrectly {
+	if util.CrdbTestBuild {
 		if numClosed := atomic.LoadInt32(f.testingInfo.numClosed); numClosed != f.testingInfo.numClosers {
 			colexecerror.InternalError(errors.AssertionFailedf("expected %d components to be closed, but found that only %d were", f.testingInfo.numClosers, numClosed))
 		}
@@ -1088,7 +1088,7 @@ func (s *vectorizedFlowCreator) setupFlow(
 			if flowCtx.EvalCtx.SessionData.TestingVectorizeInjectPanics {
 				result.Root = newPanicInjector(result.Root)
 			}
-			if flowCtx.Cfg != nil && flowCtx.Cfg.TestingKnobs.CheckVectorizedFlowIsClosedCorrectly {
+			if util.CrdbTestBuild {
 				toCloseCopy := append(colexecop.Closers{}, result.ToClose...)
 				for i := range toCloseCopy {
 					func(idx int) {

--- a/pkg/sql/execinfra/server_config.go
+++ b/pkg/sql/execinfra/server_config.go
@@ -213,10 +213,6 @@ type TestingKnobs struct {
 	// checked by a test receiver on the gateway.
 	MetadataTestLevel MetadataTestLevel
 
-	// CheckVectorizedFlowIsClosedCorrectly checks that all components in a flow
-	// were closed explicitly in flow.Cleanup.
-	CheckVectorizedFlowIsClosedCorrectly bool
-
 	// Changefeed contains testing knobs specific to the changefeed system.
 	Changefeed base.ModuleTestingKnobs
 

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1357,8 +1357,6 @@ func (t *logicTest) newCluster(serverArgs TestServerArgs) {
 
 	distSQLKnobs := &execinfra.TestingKnobs{
 		MetadataTestLevel: execinfra.Off,
-		// TODO(yuzefovich): uncomment once #64248 is resolved.
-		//CheckVectorizedFlowIsClosedCorrectly: true,
 	}
 	cfg := t.cfg
 	if cfg.sqlExecUseDisk {

